support/docker: Add build shim

samueldr opened the pull request:
This adds a shim allowing end-users to invoke a containerized Nix
instance, in a manner compatible to using `nix-build`.

```
 $ support/docker/build.sh -A pine64-pineA64LTS
```

This is equivalent to

```
 $ nix-build -A pine64-pineA64LTS
```

The semantics vary a little. *Only* the `result` symlink is resolved and
replaced with the store path contents. This means that using
`--out-link` is not supported at this time, and multiple build outputs
won't work as expected.

This should be sufficient though.

* * *

This is desirable as an *alternative known good way* to invoke the Nix build on non-NixOS. Installing Nix globally using the upstream installer should work just as well. Using the distro-provided packages is hit-and-miss, due to unclear reasons, depending on the distro.

*In addition*, it may be useful for macOS or Windows users using *whatever cursed thing docker provides for them*.

* * *

Tested on Ubuntu

![image](https://user-images.githubusercontent.com/132835/151468825-f85c7e53-9f03-4f6f-b25a-6716c80a35d3.png)


Fixes #5

samueldr committed:
support/docker: Add build shim

This adds a shim allowing end-users to invoke a containerized Nix
instance, in a manner compatible to using `nix-build`.

```
 $ support/docker/build.sh -A pine64-pineA64LTS
```

This is equivalent to

```
 $ nix-build -A pine64-pineA64LTS
```

The semantics vary a little. *Only* the `result` symlink is resolved and
replaced with the store path contents. This means that using
`--out-link` is not supported at this time, and multiple build outputs
won't work as expected.

This should be sufficient though.

Co-authored-by: Linus Heckemann <git@sphalerite.org>

lheckemann commented on support/docker/build.sh:
```diff
@@ -0,0 +1,61 @@
+	local STORE_VOLUME
+	STORE_VOLUME="$(docker volume create "Tow-Boot--nix-store")"
+
+	exec docker run \
```
Using `exec` in a function can be confusing, since that causes it not to return. That doesn't make a difference in this scenario, but if the script ever does more outside docker it could get confusing. Maybe remove that here?

samueldr replied:
Yeah, I even had the "debate" internally... I wanted to get rid of the needless parent shell process... But at what cost?

lheckemann commented on support/docker/build.sh:
```diff
@@ -0,0 +1,61 @@
+
+	exec docker run \
+		--volume "$STORE_VOLUME:/nix" \
+		--volume "$PWD:/Tow-Boot" \
```
This assumes that the user is currently in the Tow-Boot root directory.

```suggestion
		--volume "$(cd "$(dirname "$0")/../.."; pwd):/Tow-Boot" \
```
While this also makes an assumption (that the script hasn't been moved elsewhere), I think that's an assumption that's more likely to hold.

lheckemann commented on support/docker/build.sh:
```diff
@@ -0,0 +1,61 @@
+else
+	# Outside the docker environment, we're calling this script in the container.
+	_nix_docker /Tow-Boot/support/docker/build.sh \
+		"$(id -u "$USER")" "$(id -u "$USER")" \
```
```suggestion
		"$(id -u "$USER")" "$(id -g "$USER")" \
```
or potentially only
```suggestion
		"$(id -u)" "$(id -g)" \
```
It's quite unclear to me where the environment variable USER comes from (it's certainly not bash) and I think using the IDs that we're running as is more likely to behave as desired than a user database lookup.

lheckemann commented on support/docker/build.sh:
```diff
@@ -0,0 +1,61 @@
+	)
+
+	# Unwrap the `result` symlink
+	out=$(readlink -f result)
```
Capturing stdout of `nix-build` could be better than doing this, not reliant on only one derivation being built and `--no-out-link` or `-o` not being passed. I think it's also less racy?

samueldr replied:
I'm concerned about multiple outputs though. And yes I hate that part.

If you have any suggestions, I'm all ears.

lheckemann replied:
How about building outside the `/Tow-Boot` volume (so the directory where the result symlink ends up is owned purely by the container) and then copying everything that ends up in said directory to `/Tow-Boot` afterwards? This would further break passing paths (currently it behaves as one would expect if the script is invoked from the tow-boot checkout) but I'm questioning whether we should really pass the arguments to `nix-build` directly at all --- I feel like too much of the interface is broken by this script and allowing the user to pass only attrs from the top level of Tow-Boot might be better than passing everything through as is to `nix-build`.

samueldr replied:
I wonder too. A goal I tried to achieve is that it should map as closely to Nix as possible, such that upgrading to using Nix proper just needs to change the invocation to `nix-build`.

lheckemann replied:
In that case, the script should probably respect the user's cwd. Passing a path will, in the current incarnation, only work as expected if the user _happens_ to be in the root of the Tow-Boot repo.

samueldr replied:
*except* we're in the docker container, which means we can't know the actual CWD.

This could be done, though, by detecting the CWD relative to the root of the project, and only accept incantations inside the repo.

lheckemann replied:
Exactly, we're already wrapping docker in this shell script, so we can also work out where we are relative to the repo root and pass it using `docker run`'s `-w`.

samueldr replied:
I think we have to sacrifice at least one part of the CLI for this to work without parsing the whole CLI.

We *can* build in another directory to easily expand all result links, but `./.` now refers to somewhere else, and thus `error: getting status of ...: No such file or directory`.

Given I prefer expanding all out links to allow building arbitrary `.nix` files, I think the concession will be to forcibly refer to `<Tow-Boot/default.nix>`. Anyone needing something more involved is free to edit the script, install Nix or use some other means. At least *everything else* on the CLI should be fine, other than referring to additional paths. Which is fine, since I dropped supporting paths for `--arg device` in #89.

> I'm using the angled brackets here to hide the implementation detail of the volume... Otherwise `/Tow-Boot/default.nix` would be used, which at this point basically has as much value.

If you have any ideas how to have our cake and eat it too (allow arbitrary relative paths to still work), I'm all ears.

lheckemann replied:
Fair enough. This was mostly about making sure the limitations are a conscious decision and not an unfortunate accident :)

>have our cake and eat it too

We could mount the entire host system into the container. But that would be quite rude.

samueldr replied:
It wouldn't solve identifying new build outputs to unwrap! With the current strategy, the build needs to happen in an empty directory, such that we can assume `*` represents all out paths.

The issue is on the CLI `./.` is CWD, and cannot be changed, so there is no way to have an empty directory, and still having `./.` refer to somewhere else.

Which is why it now forcibly builds `<Tow-Boot/default.nix>`

lheckemann commented on support/docker/build.sh:
```diff
@@ -0,0 +1,61 @@
+	chown -R "$REAL_UID:$REAL_GID" result
+else
+	# Outside the docker environment, we're calling this script in the container.
+	_nix_docker /Tow-Boot/support/docker/build.sh \
```
Personally I like putting "environment setup" at the top of the script, checking if we're in the environment we want (`[ -e /Tow-Boot ]` in this script) and `exec`ing if not, then having the material part of the script outside any conditions. That's just a matter of taste though :)

samueldr replied:
As you said ¯\\\_(ツ)\_/¯.

In more complex situations I would implement more bash functions and at the end of the script would be the tiniest amount of code to "route" to the right functions. But this is of a relatively manageable size considering the magic it does behind the scenes.

samueldr commented:
Other than the current (1) unresolved discussion, I believe this should be ready.

At worst, the unresolved discussion would be something to consider in re-evaluating the script later down the line.

lheckemann commented on support/docker/build.sh:
```diff
@@ -0,0 +1,61 @@
+	# Unwrap the `result` symlink
+	out=$(readlink -f result)
+	rm result
+	cp -r "$out" result
+	chown -R "$REAL_UID:$REAL_GID" result
```
Will the copy always be fast enough that reporting what's happening isn't necessary?

samueldr replied:
No idea. Outputs could be big, computers slow.

lheckemann replied:
Then it should probably report that it's copying the output --- since Nix itself has finished at this point, it looks like the whole process could be complete and the outputs "ready to use".

samueldr replied:
It's now quite obvious.

![image](https://user-images.githubusercontent.com/132835/153524476-c6380231-8e59-483e-916b-97206347b469.png)

samueldr committed:
docker/build: Honor relative pwd

And thus disallow execution outside of the project directory.

samueldr committed:
docker/build: Print message about unwrapping outputs

samueldr committed:
docker/build: Unwrap all results, but force default.nix

This way we can build multiple outputs in one go, but we're forced to
abandon arbitrary Nix evaluations.

This also helps reduce rudeness with removing `result` unconditionally,
which wouldn't be helpful when given `-o foo`

samueldr committed:
doc: Add link to developer guide in getting started

samueldr committed:
doc: Improve development guide

Mainly adds resolved details about a docker build setup.

samueldr commented:
- Updated with current state of `development`
 - Made the shim honor project-relative cwd
 - Made the script unwrap all output paths
 - Updated documentation
